lib/events/stream.go

```diff
@@ -0,0 +1,694 @@
+	switch chunk.GetType() {
+	// Open the tar archive by writing the header. Since this is a raw stream
+	// the size of the content to be written is known.
+	case typeOpenRaw:
+		err := s.tarWriter.WriteHeader(&tar.Header{
+			Name: chunk.GetFileName(),
+			Mode: 0600,
+			Size: chunk.GetFileSize(),
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	// Close is a NOP because writing a header indicates the size of file and
+	// where the next file starts.
+	case typeCloseRaw:
+	// Chunk can be written directly to the tar archive.
+	case typeChunkRaw:
+		s.rawChunkCount = s.rawChunkCount + 1
+
+		_, err = s.tarWriter.Write(chunk.GetData())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// processEvents takes chunks, validates them, and then buffers them in a
+// gzip stream until complete then writes them to the tar archive.
+func (s *Stream) processEvents(chunk *proto.SessionChunk) error {
+	var err error
+
+	switch chunk.GetType() {
+	case typeOpenEvents:
+		// Take semaphore before creating zip archive in memory.
+		err = s.manager.takeSemaphore()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Get a buffer from the pool.
+		s.zipBuffer = s.manager.pool.Get().(*bytes.Buffer)
+
+		s.zipWriter, err = gzip.NewWriterLevel(s.zipBuffer, gzip.BestSpeed)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	case typeCloseEvents:
+		// Close zip file and after writing it to the tar archive, release
+		// any resources.
+		err = s.zipWriter.Close()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		s.zipWriter = nil
+		defer s.zipBuffer.Reset()
+		defer s.manager.pool.Put(s.zipBuffer)
+
+		// Copy the zip archive into the tar stream.
+		err := s.tarWriter.WriteHeader(&tar.Header{
+			Name: chunk.GetFileName(),
+			Mode: 0600,
+			Size: int64(s.zipBuffer.Len()),
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		_, err = io.Copy(s.tarWriter, s.zipBuffer)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Done processing zip archive, release semaphore.
+		err = s.manager.releaseSemaphore()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	case typeChunkEvents:
+		s.eventsChunkCount = s.eventsChunkCount + 1
+
+		// Validate incoming event.
+		var f EventFields
+		err = utils.FastUnmarshal(chunk.GetData(), &f)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		err := ValidateEvent(f, s.serverID)
+		if err != nil {
+			s.manager.log.Warnf("Rejecting audit event %v from %v: %v. A node is attempting to "+
+				"submit events for an identity other than the one on its x509 certificate.",
+				f.GetType(), s.serverID, err)
+			return trace.AccessDenied("failed to validate event")
+		}
+
+		// Write event to zip buffer.
+		_, err = s.zipWriter.Write(append(chunk.GetData(), '\n'))
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// checkTransition makes sure the archive is being created in a sane manner.
+func (s *Stream) checkTransition(chunk *proto.SessionChunk) error {
```

Comment:
Does not detect double-init or missing init.  Double-init appears to leak resources.  Missing init appears to create panics.